go/host/enclave/service.go

```diff
@@ -205,3 +220,178 @@ func (e *Service) promoteNewActiveSequencer() {
+		e.logger.Debug("No active sequencer, skipping periodic rollup.")
+		return false, 0
+	}
+
+	fromBatch, err := e.getLatestBatchNo()
+	if err != nil {
+		e.logger.Error("Encountered error while trying to retrieve latest batch", log.ErrKey, err)
+		return false, 0
+	}
+
+	// estimate the size of a compressed rollup
+	availBatchesSumSize, err := e.calculateNonRolledupBatchesSize(fromBatch)
+	if err != nil {
+		e.logger.Error("Unable to estimate the size of the current rollup", log.ErrKey, err, "from_batch", fromBatch)
+		// Note: this should not happen. If it does, we will assume the size is 0, meaning only time will trigger a rollup
+		availBatchesSumSize = 0
+	}
+
+	// adjust the availBatchesSumSize
+	estimatedRunningRollupSize := uint64(float64(availBatchesSumSize) * batchCompressionFactor)
+
+	// produce and issue rollup when either:
+	// it has passed g.rollupInterval from last lastSuccessfulRollup
+	// or the size of accumulated batches is > g.maxRollupSize
+	timeExpired := time.Since(lastSuccessfulRollup) > e.rollupInterval
+	sizeExceeded := estimatedRunningRollupSize >= e.maxRollupSize
+
+	return timeExpired || sizeExceeded, fromBatch
+}
+
+func (e *Service) prepareRollup(guardian *Guardian, fromBatch uint64) (*common.CreateRollupResult, error) {
+	enclID := guardian.GetEnclaveID()
+	e.logger.Info("Attempting to produce rollup.", log.EnclaveIDKey, enclID)
+	result, err := guardian.GetEnclaveClient().CreateRollup(context.Background(), fromBatch)
+	if err != nil {
+		e.logger.Info("Unable to produce rollup", log.EnclaveIDKey, enclID, log.ErrKey, err)
+		return nil, err
+	}
+	rollup, err := ethadapter.ReconstructRollup(result.Blobs)
+	if err != nil {
+		e.logger.Error("Failed to reconstruct rollup", log.ErrKey, err)
+		return nil, err
+	}
+
+	canonBlock, err := e.sl.L1Data().FetchBlockByHeight(rollup.Header.CompressionL1Number)
+	if err != nil {
+		e.logger.Error("Failed to fetch canonical block for rollup", log.ErrKey, err)
+		return nil, err
+	}
+
+	// only publish if the block used for compression is canonical
+	if canonBlock.Hash() != rollup.Header.CompressionL1Head {
+		e.logger.Info("Skipping rollup publication because compression block is not canonical.", "block", canonBlock.Hash())
+		return nil, fmt.Errorf("compression block is not canonical, block=%s", canonBlock.Hash())
+	}
+	return result, nil
+}
+
+func (e *Service) getLatestBatchNo() (uint64, error) {
+	lastBatchNo, err := e.sl.L1Publisher().FetchLatestSeqNo()
+	if err != nil {
+		return 0, err
+	}
+	fromBatch := lastBatchNo.Uint64()
+	if lastBatchNo.Uint64() > common.L2GenesisSeqNo {
+		fromBatch++
+	}
+	return fromBatch, nil
+}
+
+func (e *Service) calculateNonRolledupBatchesSize(seqNo uint64) (uint64, error) {
+	var size uint64
+
+	if seqNo == 0 { // don't calculate for seqNo 0 batches
+		return 0, nil
+	}
+
+	currentNo := seqNo
+	for {
+		batch, err := e.sl.L2Repo().FetchBatchBySeqNo(context.TODO(), big.NewInt(int64(currentNo)))
```

Comment:
this feels highly inefficient. And I don't think we're caching batches on the host. Basically every second, we run a lot of redundant queries. This will cause performance issues with higher rollup times. Would it be hard to keep a rolling "currentRollupSize", and append to it the incoming batches